### ISSUE TITLE
Drag rectangle doesn't take into consideration page scrolling.

### ISSUE THREAD

**Issue by jreijt** [d3lm/ngx-drag-to-select#36]
This is a bit hard to explain and I can't really provide screenshots for it, but here it goes. 

Case 1: 
Dragging while the page is scrolled to top. 
Expected behavior: Rectangle starts where mouse clicks and drags where the mouse goes.
Actual behavior: Same as expected (Y) 

Case 2:
Dragging while the page is scrolled. 
Reproduction steps: 

1. Scroll down a bit
2. Try to drag.

Expected behavior: Rectangle starts where mouse clicks and drags where the mouse goes.
Actual behavior: Rectangle starts where the mouse would have been if we hadn't scrolled down, and it drags relatively to its start position. 

So basically what I think is happening is that the start position of the rectangle when we drag isn't taking into consideration the page scrollY. 

Does that make sense? I'll be happy to provide more information.

**Comment by d3lm** [comment]
Hey @jreijt, thanks for opening up this issue!

It's a little hard to follow. Maybe you can provide a GIF or a record your screen to see what is not behaving as expected. Would be awesome if you could elaborate a little bit more on this.

Dragging while scrolling works as expected for me, if I understood you correctly.

![dragging_scrolling](https://user-images.githubusercontent.com/12571019/43200105-f3cc1c14-901c-11e8-8f76-fc123c479d1c.gif)

**Note**: The GIF tool I used somehow slowed down my system, that's why the selection seems a little laggy in the GIF.

**Comment by d3lm** [comment]
Ah I think I now got it what you mean. You mean that items are not selected while dragging and scrolling. You have to move the mouse in order to select items, is that right?

**Comment by jreijt** [comment]
Yeah GIF looks like the way to go for this one. 

So here's the first case, dragging before scrolling (which works fine). 
![case 1](https://user-images.githubusercontent.com/25050560/43200683-f8368058-901e-11e8-8c19-a8f19b394824.gif)

When I scroll down and try to drag, the start point of the rectangle doesn't fall on the mouse. 
![case2](https://user-images.githubusercontent.com/25050560/43200850-82ad9d84-901f-11e8-958e-47c1cf4fd772.gif)

This is what the HTML looks like. Basically I'm wrapping the "bag-group" component with the DTS directive. The bag-group component is the one holding all the cards. 
![screen shot 2018-07-25 at 3 32 41 pm](https://user-images.githubusercontent.com/25050560/43201068-255c5a2a-9020-11e8-9988-46ad0e156af1.png)

**Comment by d3lm** [comment]
Yes, moving the cursor to the edges of your screen will not scroll the window. This is not _yet_ implemented but planned at some point.

On your GIF, it's hard to tell what is going wrong there. It seems that the bounding rects are not correctly calculated in your case or the initial mouse position is wrong. How many levels of nesting do you have? Can you provide a minimalistic reproducable on StackBlitz so I can investigate what's going wrong? Otherwise it's more like taking a shot in the dark. I can see that you have the `select-container` instance inside an expansion component or vertical tab. Maybe what would work is when you open the "tab" you call `open` on the `select-container` which will recalculate the bounding rects for the container as well as for all selectable items.

**Comment by sam-lex** [comment]
I had this issue too, but solved by calling `SelectContainerComponent.update()` when scrolling
```html
<div (scroll)="_onScroll($event)"> <!-- hook scroll event to the div that has overflow  -->
  <dts-select-container>
```
```typescript
@ViewChild(SelectContainerComponent) dtsContainer: SelectContainerComponent;

_onScroll(event) {
  // I send it to a "scroll observable" to use it with debounceTime( 5 )
  this.dtsContainer.update();
}
```

Dragging + Scrolling at the same time works well too...
If you do use `debounceTime`, you might see a slight delay for the selection box to "catch up" to your mouse cursor when scrolling. It was necessary for me to use it because I needed different divs to sync their scrollbars, and without it, the scrollbars would get the wrong values...  
So maybe it will be needed when updating the container too, I'm not sure

**Comment by d3lm** [comment]
@sam-lex Did you have the same problem that when dragging and scrolling items were not selected, or what do you mean?

I could think of one solution that would be quite generic and it could tackle 90% of the apps. It would be to listen for scroll events on the `body` and trigger the selection logic here too. But this would not solve the problem of having multiple nested scrollable containers and this is something that varies a lot between the apps. I am not sure from the top of my head how to solve this one because it requires to take into account the scroll position of each and every scrollable container.

**Comment by jreijt** [comment]
@sam-lex Thank you so much! That's exactly what I needed. 

@d3lm  The update on scroll did indeed the problem I was having with the rectangle location. To have it be a little more specific I can also make it so only the `select-container` that's expanded updates on scroll. I'm thinking this should be mentioned somewhere on your page though. A lot of cases I can think of in Angular where I'd use drag to select have some kind of nested view like this, like in my case with components nested inside card-group components, nested inside the main page.

Anyway, thanks again for your help gents. You can close this issue.

**Comment by sam-lex** [comment]
Yes, I had a my `select-container` inside a scrollable div, and the results were exactly the same as @jreijt 2nd Gif.  
basically from what I understand is this:
you scroll down say `50px`  
start select-drag at `(100, 100 ) px`  
The rectangle box will start drawing from `(100, 100)`  
but since you scrolled down `50px`, it should draw from `(100, 150)px`

**Comment by sam-lex** [comment]
I don't know how this could be solved gracefully through the lib though, take this scenario for example:  
nested scrollable div with `height: 30vh`  
you start select-dragging, and move past the edge of your nested-div, what should happen?  
- auto-scroll?
- if not auto-scroll, then should the selectBox stop at the edge? or continue through "inside" what's overflown?

**Comment by d3lm** [comment]
@jreijt Totally, this needs to go into the docs. Very good point! The lib only accounts for the body to be scrolled (document) and the `dts-select-container` itself. For any other element that is scrollable which wraps the select-container, the consumer needs to take care of calling `update` which will recalculate the positions of all selectable items and the select container itself. This is required in order to render out the correct select-box.

@sam-lex Yes this is not super trivial to solve. Do you think this is something that is needed? I thought it would be cool to have this. One way is to introduce another input to enable auto-scroll. If set to true the window edges serve as the controller for scrolling. If you move your cursor to the bottom edge the window will scroll. But then again, what happens if it's not the window but a nested scrollable container that you want to scroll. We could have another input for `auto-scroll-container` which is then scrolled if `auto-scroll` is `true`. Just thinking out loud here. But this would introduce some logic to the lib making it bigger. I want to keep it as lightweight as possible. So if we introduce something like this we would need a few people that ask for this feature I guess. What do you think?

**Comment by d3lm** [comment]
Closing this as it can be solved with the APIs available and there will be an info in the docs once #37 is implemented.

**Comment by d3lm** [comment]
BTW, for what it's worth, it's super cool to see what you folks are using this lib for. This makes me extremely happy. Thanks a bunch for using it for your apps 👍🙌

**Comment by sam-lex** [comment]
@d3lm I like the idea of using inputs, it keeps the component flexible and customizable.  
I don't know the solution for this one either, but here's an idea:  
`@Input hostContainer` (name to be determined)  
when this input is defined, the lib will use this container to base its mouse/scroll positions

if this works, I think it'd be necessary to solve the issue of "dragging to the edge", should it stop or continue? or maybe have another input for this?  
Personally I think it should stop at the edge if auto-scroll is disabled, otherwise it would select things that are not visible to the user, but with so many use cases out there, an Input to override it might be good.

**Comment by d3lm** [comment]
I would say it doesn't stop at the edges. It just continues until the `hostContainer` cannot be scrolled any further. Just thinking out loud here.

**Comment by d3lm** [comment]
@jreijt @sam-lex Would be great to get your feedback on this [PR](https://github.com/d3lm/ngx-drag-to-select/pull/38) for the FAQ section. Does this answer the question on how to deal with this scenario?

**Comment by jreijt** [comment]
@d3lm Looks good mate. Great work on the lib btw.

**Comment by d3lm** [comment]
Thanks @jreijt, means a lot to me!

**Comment by DawsonLey** [comment]
I'm calling update on the select-container when the page scrolls, but no luck, select box is still offset from mouse by the amount I scrolled down. Am I missing something? The method is definitely getting called, it's just not working :/
